sys/arm64/arm64/machdep.c

```diff
@@ -199,7 +203,12 @@ fill_regs(struct thread *td, struct reg *regs)
 	regs->elr = frame->tf_elr;
 	regs->spsr = frame->tf_spsr;
 
+#if !__has_feature(capabilities)
+	for (int i = 0; i < nitems(regs->x); i++)
+		regs->x[i] = mcp->mc_gpregs.gp_x[i];
+#else
 	memcpy(regs->x, frame->tf_x, sizeof(regs->x));
+#endif
```

Comment:
Why is it again we're not using `memcpy`?

Reply:
Because `struct reg` has an array of 64-bit GPRs.  This is what we do for now and what we will always have to do for hybrid.  Eventually for CheriABI I would like the native `struct reg` on architectures with merged register files to really be `struct capreg`.  However, that's a long term project and not something to do anytime soon.

Reply:
Ah right, yeah. But the mcontext ones below should be memcpy'ing and using `mc_capregs`?

Reply:
We could just always do the loop? I'd be extremely surprised if the compiler can't turn that into a memcpy in the sequential case.

Reply:
When we have purecap support the loop will be removed. When building with `-ffreestanding` the compiler may not convert the loop to a memcpy (older compilers didn't, but I haven't tried with a modern clang)

Reply:
Yeah, I guess -ffreestanding would prevent that optimization :(

Reply:
Even with purecap the loop is stuck for now because struct reg is always GPRs currently.

Reply:
Which is because this PR is trying to land an incomplete set of changes that don't make sense in isolation; if the headers were updated to match the assumptions made throughout the kernel then this loop wouldn't be needed. I would much rather see that patch come along with this set so we don't commit workarounds that will be immediately reverted, it needlessly pollutes and obfuscates the history.

Reply:
Eh, to be clear, this code (fill_regs) won't change even in a purecap kernel or with better CheriABI support.  The *mcontext routines might change, but not this one.  Changing the layout of `struct reg` for purecap is something I'd like to do eventually for all CHERI platforms, but that is a future change.

Reply:
Oh I got confused which one we were in (the outdated diff talking about `mc_gpregs` doesn't help). I was talking about the *mcontext ones, ie the ones that use `memcpy` on RISC-V.